### ISSUE TITLE
Deprecate IFinishableConsumer so it can be deleted

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#2661]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#2661|
|Type|task|
|Created|2007-05-18 18:04:58Z|
|Branch|https://github.com/twisted/twisted/tree/deprecate-ifinishable-consumer-2661-2|

```IFinishableConsumer``` offers no advantages over ```IConsumer```.  ```IFinishableConsumer.finish``` is effectively identical to ```IConsumer.unregisterProducer```.

<details><summary>Searchable metadata</summary>

```
trac-id__2661 2661
type__task task
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_deprecate_ifinishable_consumer_2661_2 branches/deprecate-ifinishable-consumer-2661-2
branch_author__cyli__thijs cyli, thijs
status__closed closed
resolution__fixed fixed
component__core core
keywords__easy easy
time__1179511498000000 1179511498000000
changetime__1305655240000000 1305655240000000
version__None None
owner__thijs thijs
cc__thijs
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="Peaker's avatar" src="https://avatars.githubusercontent.com/u/48899?s=50" width="50" height="50">](https://github.com/Peaker)<a name="note_1"></a>|@Peaker commented|
|-|-|

I think that if it is deleted/deprecated, IConsumer should, in addition to unregisterProducer, contain a finish() method.

While unregisterProducer() means "I will not produce any more data for you, but someone else may", finish() would mean "Nobody will ever produce any more data for you, please terminate your consumption and clean up".

This could map to meaningful operations on various consumers.
For example, a Process is a consumer that can map "finish" to closeChildFD(0) (loseConnection also closes other process pipes and is not the correct operation when consumption is complete).

Then, to map this extended Consumer to a Protocol via an adapter (as is done in twisted/internet/protocol.py) you would map finish to writeConnectionLost.

A Process would map (as it does today) all consumer methods to childfd(0). Being a FileDescriptor, it would map finish() to a loseWriteConnection, which in the case of a ProcessWriter would convert to a loseConnection.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra set owner to @cyli|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli commented|
|-|-|

(In [29624]) Branching to 'deprecate-ifinishable-consumer-2661'

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_5"></a>|@cyli commented|
|-|-|

Took Peakness's comment and opened a new ticket [#4554](https://github.com/twisted/twisted/issues/4554), which should either be closed as won't-fix or in some way validated.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_6"></a>|@cyli commented|
|-|-|

Er,  Peaker, sorry.

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_7"></a>|@cyli commented|
|-|-|

(In [29625]) Deprecate IFinishableConsumer, and test that it is indeed deprecated.

refs [#2661](https://github.com/twisted/twisted/issues/2661)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_8"></a>|@cyli removed owner|
|-|-|

[Build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/deprecate-ifinishable-consumer-2661)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun set owner to @cyli|
|-|-|

We should get rid of the one use of `IFinishableConsumer` in Twisted before putting this deprecation in place.  This will both avoid introducing new deprecation warnings emitted from Twisted's internal uses of deprecated APIs and will demonstrate that we really don't need the interface. :)

`twisted.protocols.ftp.SenderProtocol` is the one implementation we have.  I filed [#4580](https://github.com/twisted/twisted/issues/4580) for this.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_11"></a>|@thijstriemstra commented|
|-|-|

(In [31635]) update version, refs [#2661](https://github.com/twisted/twisted/issues/2661)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

(In [31636]) Branching to 'deprecate-ifinishable-consumer-2661-2'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_13"></a>|@thijstriemstra commented|
|-|-|

(In [31637]) merge forward, resolve conflicts, and update version nr. refs [#2661](https://github.com/twisted/twisted/issues/2661)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_14"></a>|@thijstriemstra commented|
|-|-|

I merged the changes, here are the build [results](http://buildbot.twistedmatrix.com/builders/lucid64-py2.6-poll/builds/735).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_16"></a>|@exarkun set owner to @cyli|
|-|-|

Thanks!

  1. The test method docstrings in `twisted.internet.test.test_interfaces.py` should drop the _Ensure_ prefix.  Also I don't think these two things are really testing different cases, but since the tests are already written, whatever.
  1. `lookForDeprecationWarning` should have a docstring.
  1. The news fragment doesn't follow the [ReviewProcess#Newsfiles style guide].  Replace _was_ with _is now_

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_17"></a>|@thijstriemstra commented|
|-|-|

(In [31837]) Address review comments, refs [#2661](https://github.com/twisted/twisted/issues/2661)

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)|@thijstriemstra removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jonathanj's avatar" src="https://avatars.githubusercontent.com/u/112166?s=50" width="50" height="50">](https://github.com/jonathanj)<a name="note_19"></a>|@jonathanj set owner to @thijstriemstra|
|-|-|

Looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_20"></a>|@thijstriemstra set status to `closed`|
|-|-|

(In [31908]) Merge deprecate-ifinishable-consumer-2661-2: twisted.internet.interfaces.IFinishableConsumer is now deprecated, use IConsumer instead. 

Author: cyli, thijs
Reviewer: exarkun, jonathanj
Fixes: [#2661](https://github.com/twisted/twisted/issues/2661)
